src/core/tUnification.ml

```diff
@@ -968,37 +976,54 @@ and unifies_to_field uctx a b ab tl (t,cf) =
 		| _ -> die "" __LOC__)
 
 and unify_with_variance uctx f t1 t2 =
-	let allows_variance_to t tf = type_iseq uctx tf t in
-	match follow t1,follow t2 with
+	let t1 = follow t1 in
+	let t2 = follow t2 in
+	let unify_param t1 t2 = with_variance (get_nested_context uctx) f t1 t2 in
+	let get_after_abstract_context () = get_after_abstract_context uctx t1 t2 in
+	let rec get_underlying_type t = match follow t with
+		| TAbstract(a,tl) ->
+			let tl = List.map get_underlying_type tl in
+			let t = apply_params a.a_params tl a.a_this in
+			if Meta.has Meta.CoreType a.a_meta then t else get_underlying_type t
+		| t -> t
+	in
+	let compare_underlying equality_kind a b = type_eq {uctx with equality_kind = equality_kind} a b in
+	let unifies_abstract uctx t ab tl ats =
+		List.exists (does_func_unify_arg (fun at ->
+			let at = apply_params ab.a_params tl at in
+			if ats == ab.a_to then
+				with_variance uctx f at t
+			else
+				with_variance uctx f t at
+		)) ats
+	in
+	let fail () = error [cannot_unify t1 t2] in
+	match t1,t2 with
 	| TInst(c1,tl1),TInst(c2,tl2) when c1 == c2 ->
-		List.iter2 f tl1 tl2
+		List.iter2 unify_param tl1 tl2
 	| TEnum(en1,tl1),TEnum(en2,tl2) when en1 == en2 ->
-		List.iter2 f tl1 tl2
-	| TAbstract(a1,tl1),TAbstract(a2,tl2) when a1 == a2 && Meta.has Meta.CoreType a1.a_meta ->
-		List.iter2 f tl1 tl2
-	| TAbstract(a1,pl1),TAbstract(a2,pl2) ->
-		if (Meta.has Meta.CoreType a1.a_meta) && (Meta.has Meta.CoreType a2.a_meta) then begin
-			let ta1 = apply_params a1.a_params pl1 a1.a_this in
-			let ta2 = apply_params a2.a_params pl2 a2.a_this in
-			type_eq {uctx with equality_kind = EqStrict} ta1 ta2;
-		end;
-		if not (List.exists (allows_variance_to t2) a1.a_to) && not (List.exists (allows_variance_to t1) a2.a_from) then
-			error [cannot_unify t1 t2]
-	| TAbstract(a,pl),t ->
-		type_eq {uctx with equality_kind = EqBothDynamic} (apply_params a.a_params pl a.a_this) t;
-		if not (List.exists (fun t2 -> allows_variance_to t (apply_params a.a_params pl t2)) a.a_to) then error [cannot_unify t1 t2]
-	| t,TAbstract(a,pl) ->
-		type_eq {uctx with equality_kind = EqBothDynamic} t (apply_params a.a_params pl a.a_this);
-		if not (List.exists (fun t2 -> allows_variance_to t (apply_params a.a_params pl t2)) a.a_from) then error [cannot_unify t1 t2]
-	| (TAnon a1 as t1), (TAnon a2 as t2) ->
-		rec_stack unify_stack (t1,t2)
-			(fun (a,b) -> fast_eq a t1 && fast_eq b t2)
-			(fun() -> unify_anons uctx t1 t2 a1 a2)
-			(fun l -> error l)
+		List.iter2 unify_param tl1 tl2
+	| TAbstract(a1,tl1),TAbstract(a2,tl2) when a1 == a2 ->
+		List.iter2 unify_param tl1 tl2
+	| TAbstract(a1,tl1),TAbstract(a2,tl2) ->
+		let uctx = get_after_abstract_context() in
+		compare_underlying EqStrict (get_underlying_type t1) (get_underlying_type t2);
```

Comment:
Looking at this I'm wondering why we do something with the underlying type here. I realize that the previous implementation already does this as well, but I can't immediately tell why.

Reply:
I guess there might be two reasons:

1) For static targets physical difference matters and so if you have `Array<Int>` you can't simply cast it to `Array<Float>`, even though `Int` defines `to Float`. Though allowed cast from/to dynamic in abstract-t/t-abstract contradicts this (if i understand it right).

2) If you allow an abstract over a child with defined cast to a parent to unify with a parent in a type param, this will permit unchecked covariance.

Reply:
But `Array<Int>` to `Array<Float>` has nothing to do with underlying types. Array is a normal class.

Shouldn't all this be covered by the `to` and `from` declarations? I don't see why the relationship of underlying types is significant here.

Reply:
About `<Int>` to `<Float>` - issue: #2844, commit: 6c22897f89bffc4ee4b5ed3ba04fcd94cb5359b3. If underlying type checks are removed, only tests related to this fail.

But if we go only by tests, then underlying type checks can be skipped for t-abstract and abstract-t.

Reply:
Added commit with removal of underlying checks for abstract/non-abstract.

Reply:
Removed that commit. Instead added assertions showcasing second reason (so first reason covers abstract-abstract, second one - abstract-t). Example here:
```haxe
class Parent {}
class Child extends Parent {}
abstract MyChild(Child) to Parent {}

class Main {
  public static function main() {
    ((null: Array<Child>): Array<Parent>); // fails as expected
    ((null: Array<MyChild>): Array<Parent>); // works if no underlying type checks are made
  }
}
```

Reply:
I think the correct behavior here is to require bidirectional unification for type-parameters. This will then work when the abstract has `to Parent from Parent`, in which case there's no variance issue.

And then if we ever allow variance annotations on type parameters, we would only have to check one of the directions instead of both.

Reply:
What about `<Array>` to `<ReadOnlyArray>`?

Reply:
I'm not sure, that looks like it shouldn't work without `to Array<T>`, but at the same time I don't see what harm it could do.

Reply:
I think those breaking changes will require more discussion and input...
In the meantime, i would prefer to fix bugs in current behaviour (and express it in tests).
Are there some considerations that prevent this from being merged?
Performance or complexity of code?